### ISSUE TITLE
High CPU utilization due to invalid CDS configuration

### ISSUE THREAD

**Issue by fredstanley** [openservicemesh/osm#4224]
**Bug description**: OSM-controller high CPU utilization

**Affected area (please mark with X where applicable)**:
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- Demo                       [ ]
- CI System                  [ ]

**Expected behavior**:

**Steps to reproduce the bug (as precisely as possible)**:
pprof output
/ # go tool pprof -top http://127.0.0.1:2500/debug/pprof/profile?debug=1
Fetching profile over HTTP from http://127.0.0.1:2500/debug/pprof/profile?debug=1
Saved profile in /root/pprof/pprof.osm-controller.samples.cpu.001.pb.gz
File: osm-controller
Type: cpu
Time: Oct 6, 2021 at 11:22pm (UTC)
Duration: 30.12s, Total samples = 23.19s (76.99%)
Showing nodes accounting for 19.71s, 84.99% of 23.19s total
Dropped 442 nodes (cum <= 0.12s)
      flat  flat%   sum%        cum   cum%
     4.05s 17.46% 17.46%      4.05s 17.46%  sync.(*RWMutex).RLock (inline)
     3.05s 13.15% 30.62%      3.24s 13.97%  sync.(*RWMutex).RUnlock
     1.23s  5.30% 35.92%      1.51s  6.51%  runtime.mapaccess2_faststr
     0.83s  3.58% 39.50%      9.77s 42.13%  k8s.io/client-go/tools/cache.(*threadSafeMap).Get
     0.76s  3.28% 42.78%      0.76s  3.28%  runtime.futex
     0.59s  2.54% 45.32%      0.80s  3.45%  runtime.mapiternext
     0.59s  2.54% 47.87%      1.11s  4.79%  runtime.scanobject
     0.53s  2.29% 50.15%      0.66s  2.85%  runtime.heapBitsSetType
     0.52s  2.24% 52.39%      0.60s  2.59%  runtime.(*itabTableType).find
     0.49s  2.11% 54.51%      0.49s  2.11%  runtime.asyncPreempt
     0.48s  2.07% 56.58%      0.48s  2.07%  memeqbody
     0.42s  1.81% 58.39%      2.69s 11.60%  runtime.mallocgc
     0.35s  1.51% 59.90%      0.37s  1.60%  syscall.Syscall
     0.32s  1.38% 61.28%     13.45s 58.00%  github.com/openservicemesh/osm/pkg/kubernetes.Client.ListPods
     0.32s  1.38% 62.66%      0.40s  1.72%  runtime.findObject
     0.32s  1.38% 64.04%      0.32s  1.38%  runtime.memclrNoHeapPointers
     0.26s  1.12% 65.16%      0.26s  1.12%  runtime.nextFreeFast (inline)
     0.25s  1.08% 66.24%      0.86s  3.71%  runtime.getitab
     0.23s  0.99% 67.23%      1.67s  7.20%  k8s.io/client-go/tools/cache.(*threadSafeMap).List
     0.23s  0.99% 68.22%      0.23s  0.99%  runtime.memmove
     0.21s  0.91% 69.12%      1.02s  4.40%  runtime.convI2I
     0.20s  0.86% 69.99%      1.22s  5.26%  k8s.io/client-go/tools/cache.(*sharedIndexInformer).GetStore
     0.19s  0.82% 70.81%      1.39s  5.99%  runtime.growslice
     0.19s  0.82% 71.63%      0.21s  0.91%  runtime.mapaccess1_faststr
     0.18s  0.78% 72.40%      1.69s  7.29%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).getAllowedDirectionalServiceAccounts
     0.16s  0.69% 73.09%     11.38s 49.07%  github.com/openservicemesh/osm/pkg/kubernetes.Client.IsMonitoredNamespace
     0.16s  0.69% 73.78%      0.16s  0.69%  runtime.markBits.isMarked (inline)
     0.16s  0.69% 74.47%      0.16s  0.69%  runtime.usleep
     0.13s  0.56% 75.03%     14.01s 60.41%  github.com/openservicemesh/osm/pkg/endpoint/providers/kube.Client.ListEndpointsForIdentity
     0.13s  0.56% 75.59%      9.90s 42.69%  k8s.io/client-go/tools/cache.(*cache).GetByKey
     0.12s  0.52% 76.11%      1.45s  6.25%  runtime.findrunnable
     0.12s  0.52% 76.63%      0.13s  0.56%  runtime.heapBitsForAddr (inline)
     0.12s  0.52% 77.15%      0.32s  1.38%  runtime.sweepone
     0.11s  0.47% 77.62%      0.27s  1.16%  runtime.greyobject
     0.09s  0.39% 78.01%      0.24s  1.03%  runtime.checkTimers
     0.09s  0.39% 78.40%      0.17s  0.73%  runtime.lock2
     0.09s  0.39% 78.78%      0.13s  0.56%  runtime.nanotime (inline)
     0.09s  0.39% 79.17%      0.17s  0.73%  runtime.typehash
     0.08s  0.34% 79.52%      0.45s  1.94%  k8s.io/apimachinery/pkg/labels.SelectorFromValidatedSet
     0.08s  0.34% 79.86%      0.25s  1.08%  runtime.runqgrab
     0.07s   0.3% 80.16%      0.12s  0.52%  runtime.(*spanSet).push
     0.07s   0.3% 80.47%      1.77s  7.63%  runtime.schedule
     0.06s  0.26% 80.72%      1.29s  5.56%  runtime.gcDrain
     0.06s  0.26% 80.98%      0.20s  0.86%  runtime.mapaccess2
     0.05s  0.22% 81.20%      0.53s  2.29%  runtime.(*mcentral).cacheSpan
     0.05s  0.22% 81.41%      0.17s  0.73%  runtime.(*mheap).allocSpan
     0.05s  0.22% 81.63%      0.21s  0.91%  runtime.(*mspan).sweep
     0.05s  0.22% 81.85%      0.22s  0.95%  runtime.nilinterhash
     0.04s  0.17% 82.02%      1.11s  4.79%  github.com/openservicemesh/osm/pkg/endpoint/providers/kube.(*Client).getServicesByLabels
     0.04s  0.17% 82.19%      0.17s  0.73%  github.com/openservicemesh/osm/pkg/endpoint/providers/kube.Client.ListEndpointsForService
     0.04s  0.17% 82.36%      0.89s  3.84%  runtime.makeslice
     0.03s  0.13% 82.49%     14.12s 60.89%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).listEndpointsForServiceIdentity
     0.03s  0.13% 82.62%      0.65s  2.80%  github.com/openservicemesh/osm/pkg/smi.(*client).ListTrafficTargets
     0.03s  0.13% 82.75%      0.70s  3.02%  runtime.(*mcache).nextFree
     0.03s  0.13% 82.88%      0.13s  0.56%  runtime.gcWriteBarrier
     0.03s  0.13% 83.01%      0.25s  1.08%  runtime.mapassign
     0.03s  0.13% 83.14%      0.32s  1.38%  runtime.newobject
     0.02s 0.086% 83.23%      0.39s  1.68%  crypto/tls.(*Conn).Write
     0.02s 0.086% 83.31%      0.19s  0.82%  fmt.(*pp).doPrintf
     0.02s 0.086% 83.40%      0.15s  0.65%  fmt.(*pp).printArg
     0.02s 0.086% 83.48%      1.69s  7.29%  k8s.io/client-go/tools/cache.(*cache).List
     0.02s 0.086% 83.57%      0.13s  0.56%  net.ParseIP
     0.02s 0.086% 83.66%      0.31s  1.34%  runtime.markroot
     0.02s 0.086% 83.74%      0.27s  1.16%  runtime.wakep
     0.02s 0.086% 83.83%      0.13s  0.56%  runtime.wbBufFlush1
     0.02s 0.086% 83.92%      0.17s  0.73%  strings.genSplit
     0.01s 0.043% 83.96%      0.32s  1.38%  fmt.Sprintf
     0.01s 0.043% 84.00%      0.20s  0.86%  github.com/golang/protobuf/ptypes.MarshalAny
     0.01s 0.043% 84.04%     15.99s 68.95%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).ListAllowedEndpointsForService
     0.01s 0.043% 84.09%      0.16s  0.69%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).listEndpointsForService
     0.01s 0.043% 84.13%      1.62s  6.99%  github.com/openservicemesh/osm/pkg/endpoint/providers/kube.Client.GetServicesForServiceAccount
     0.01s 0.043% 84.17%      0.17s  0.73%  github.com/openservicemesh/osm/pkg/envoy/ads.respondToRequest
     0.01s 0.043% 84.22%     18.06s 77.88%  github.com/openservicemesh/osm/pkg/envoy/eds.NewResponse
     0.01s 0.043% 84.26%      0.16s  0.69%  github.com/openservicemesh/osm/pkg/envoy/eds.NewWSEdgePodClusterLoadAssignment.func1
     0.01s 0.043% 84.30%     17.72s 76.41%  github.com/openservicemesh/osm/pkg/envoy/eds.getEndpointsForProxy
     0.01s 0.043% 84.35%      0.42s  1.81%  github.com/openservicemesh/osm/pkg/kubernetes.Client.ListServices
     0.01s 0.043% 84.39%      0.18s  0.78%  google.golang.org/protobuf/proto.MarshalOptions.marshal
     0.01s 0.043% 84.43%      0.16s  0.69%  runtime.(*mcentral).uncacheSpan
     0.01s 0.043% 84.48%      0.35s  1.51%  runtime.(*mheap).alloc
     0.01s 0.043% 84.52%      0.16s  0.69%  runtime.(*mheap).alloc.func1
     0.01s 0.043% 84.56%      0.13s  0.56%  runtime.chanrecv2
     0.01s 0.043% 84.61%      0.18s  0.78%  runtime.convTstring
     0.01s 0.043% 84.65%      0.51s  2.20%  runtime.futexsleep
     0.01s 0.043% 84.69%      0.30s  1.29%  runtime.gcAssistAlloc1
     0.01s 0.043% 84.73%      0.29s  1.25%  runtime.gcDrainN
     0.01s 0.043% 84.78%      0.19s  0.82%  runtime.gopreempt_m
     0.01s 0.043% 84.82%      1.51s  6.51%  runtime.park_m
     0.01s 0.043% 84.86%      0.26s  1.12%  runtime.runqsteal
     0.01s 0.043% 84.91%      0.28s  1.21%  runtime.startm
     0.01s 0.043% 84.95%      0.60s  2.59%  runtime.stopm
     0.01s 0.043% 84.99%      0.12s  0.52%  runtime.typedmemmove
         0     0% 84.99%      0.32s  1.38%  crypto/tls.(*Conn).write
         0     0% 84.99%      0.37s  1.60%  crypto/tls.(*Conn).writeRecordLocked
         0     0% 84.99%      0.42s  1.81%  github.com/deckarep/golang-set.(*threadSafeSet).Add
         0     0% 84.99%      0.40s  1.72%  github.com/deckarep/golang-set.(*threadUnsafeSet).Add (inline)
         0     0% 84.99%      0.19s  0.82%  github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3._AggregatedDiscoveryService_StreamAggregatedResources_Handler
         0     0% 84.99%      0.19s  0.82%  github.com/golang/protobuf/proto.Marshal (inline)
         0     0% 84.99%      0.19s  0.82%  github.com/golang/protobuf/proto.marshalAppend
         0     0% 84.99%      1.69s  7.29%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).ListAllowedOutboundServiceIdentities (inline)
         0     0% 84.99%      1.71s  7.37%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).ListAllowedOutboundServicesForIdentity
         0     0% 84.99%      1.67s  7.20%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).getServicesForServiceAccount
         0     0% 84.99%      0.15s  0.65%  github.com/openservicemesh/osm/pkg/catalog.(*MeshCatalog).witesandHttpClient
         0     0% 84.99%      0.19s  0.82%  github.com/openservicemesh/osm/pkg/envoy/ads.(*Server).StreamAggregatedResources
         0     0% 84.99%     18.38s 79.26%  github.com/openservicemesh/osm/pkg/envoy/ads.(*Server).newAggregatedDiscoveryResponse
         0     0% 84.99%     18.41s 79.39%  github.com/openservicemesh/osm/pkg/envoy/ads.(*Server).sendResponse
         0     0% 84.99%     18.41s 79.39%  github.com/openservicemesh/osm/pkg/envoy/ads.(*Server).sendTypeResponse
         0     0% 84.99%     18.41s 79.39%  github.com/openservicemesh/osm/pkg/envoy/ads.(*proxyResponseJob).Run
         0     0% 84.99%      0.12s  0.52%  github.com/openservicemesh/osm/pkg/envoy/ads.getRequestedResourceNamesSet
         0     0% 84.99%      0.21s  0.91%  github.com/openservicemesh/osm/pkg/envoy/eds.NewWSEdgePodClusterLoadAssignment
         0     0% 84.99%      0.27s  1.16%  github.com/openservicemesh/osm/pkg/identity.K8sServiceAccount.ToServiceIdentity (inline)
         0     0% 84.99%      0.17s  0.73%  github.com/openservicemesh/osm/pkg/identity.ServiceIdentity.ToK8sServiceAccount
         0     0% 84.99%      0.12s  0.52%  github.com/openservicemesh/osm/pkg/service.MeshService.String (inline)
         0     0% 84.99%     18.45s 79.56%  github.com/openservicemesh/osm/pkg/workerpool.(*worker).work
         0     0% 84.99%      0.12s  0.52%  golang.org/x/net/http2.(*Framer).ReadFrame
         0     0% 84.99%      0.17s  0.73%  google.golang.org/grpc.(*Server).handleRawConn.func1
         0     0% 84.99%      0.19s  0.82%  google.golang.org/grpc.(*Server).handleStream
         0     0% 84.99%      0.19s  0.82%  google.golang.org/grpc.(*Server).processStreamingRPC
         0     0% 84.99%      0.17s  0.73%  google.golang.org/grpc.(*Server).serveStreams
         0     0% 84.99%      0.19s  0.82%  google.golang.org/grpc.(*Server).serveStreams.func1.2
         0     0% 84.99%      0.38s  1.64%  google.golang.org/grpc/internal/transport.(*bufWriter).Flush
         0     0% 84.99%      0.17s  0.73%  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams
         0     0% 84.99%      0.47s  2.03%  google.golang.org/grpc/internal/transport.(*loopyWriter).run
         0     0% 84.99%      0.47s  2.03%  google.golang.org/grpc/internal/transport.newHTTP2Server.func2
         0     0% 84.99%      0.18s  0.78%  google.golang.org/protobuf/proto.MarshalOptions.MarshalAppend
         0     0% 84.99%      0.32s  1.38%  internal/poll.(*FD).Write
         0     0% 84.99%      0.37s  1.60%  internal/poll.ignoringEINTRIO (inline)
         0     0% 84.99%      0.45s  1.94%  k8s.io/apimachinery/pkg/labels.SelectorFromSet (inline)
         0     0% 84.99%      0.45s  1.94%  k8s.io/apimachinery/pkg/labels.Set.AsSelector (inline)
         0     0% 84.99%      0.15s  0.65%  k8s.io/client-go/kubernetes/typed/core/v1.(*pods).List
         0     0% 84.99%      0.32s  1.38%  net.(*conn).Write
         0     0% 84.99%      0.32s  1.38%  net.(*netFD).Write
         0     0% 84.99%      0.64s  2.76%  runtime.(*mcache).refill
         0     0% 84.99%      0.40s  1.72%  runtime.(*mcentral).grow
         0     0% 84.99%      0.32s  1.38%  runtime.bgsweep
         0     0% 84.99%      0.12s  0.52%  runtime.chanrecv
         0     0% 84.99%      0.17s  0.73%  runtime.convT2E
         0     0% 84.99%      0.26s  1.12%  runtime.futexwakeup
         0     0% 84.99%      0.35s  1.51%  runtime.gcAssistAlloc
         0     0% 84.99%      0.30s  1.29%  runtime.gcAssistAlloc.func1
         0     0% 84.99%      1.37s  5.91%  runtime.gcBgMarkWorker
         0     0% 84.99%      1.29s  5.56%  runtime.gcBgMarkWorker.func2
         0     0% 84.99%      0.12s  0.52%  runtime.gcMarkDone
         0     0% 84.99%      0.22s  0.95%  runtime.gentraceback
         0     0% 84.99%      0.34s  1.47%  runtime.goschedImpl
         0     0% 84.99%      0.16s  0.69%  runtime.gosched_m
         0     0% 84.99%      0.17s  0.73%  runtime.lock (partial-inline)
         0     0% 84.99%      0.17s  0.73%  runtime.lockWithRank (inline)
         0     0% 84.99%      0.48s  2.07%  runtime.mPark
         0     0% 84.99%      0.23s  0.99%  runtime.markroot.func1
         0     0% 84.99%      1.69s  7.29%  runtime.mcall
         0     0% 84.99%      0.16s  0.69%  runtime.morestack
         0     0% 84.99%      0.14s   0.6%  runtime.newproc
         0     0% 84.99%      0.14s   0.6%  runtime.newproc.func1
         0     0% 84.99%      0.23s  0.99%  runtime.newstack
         0     0% 84.99%      0.46s  1.98%  runtime.notesleep
         0     0% 84.99%      0.26s  1.12%  runtime.notewakeup
         0     0% 84.99%      0.16s  0.69%  runtime.scanstack
         0     0% 84.99%      2.43s 10.48%  runtime.systemstack
         0     0% 84.99%      0.13s  0.56%  runtime.wbBufFlush
         0     0% 84.99%      0.13s  0.56%  runtime.wbBufFlush.func1
         0     0% 84.99%      0.17s  0.73%  strings.Split (inline)
         0     0% 84.99%      0.32s  1.38%  syscall.Write (inline)
         0     0% 84.99%      0.32s  1.38%  syscall.write
/ #

**How was OSM installed?**:

**Anything else we need to know?**:

**Bug report archive**:

**Environment**:
- OSM version (use `osm version`): osmv0.9.1
- Kubernetes version (use `kubectl version`):
- Size of cluster (number of worker nodes in the cluster): 3 worker nodes
- Others:

**Comment by fredstanley** [comment]
[pprof.txt](https://github.com/openservicemesh/osm/files/7297665/pprof.txt)

**Comment by shashankram** [comment]
> What version of OSM do you see this problem? v0.9.1 How did you determine that the high CPU utilization was due to large number of connected proxies?
> 
> The top on the node was showing 150% cpu utilization for osm-controller. I have around 20-25 envoy proxies running along with services.
> 
> How many replicas of osm-controller are present? Is auto scaling enabled? only one replica of osm-controller with auto scalling off.

Thanks @fredstanley. How many app pods do you have that are being managed by the single osm-controller instance?

**Comment by fredstanley** [comment]
[top.txt](https://github.com/openservicemesh/osm/files/7297731/top.txt)

**Comment by shashankram** [comment]
> ```
> 4.05s 17.46% 17.46% 4.05s 17.46% sync.(*RWMutex).RLock (inline)
> 3.05s 13.15% 30.62% 3.24s 13.97% sync.(*RWMutex).RUnlock
> 1.23s 5.30% 35.92% 1.51s 6.51% runtime.mapaccess2_faststr
> ```
> 
> The trace looks relatively similar to #1958, which we already know about, however doesn't explain the constant cpu consumption. One explanation that comes to my head could be OSM can't realize or push (for some reason) the proxy configurations, in which case I believe will be forever retried in a rather tight loop.

Except that this is with 30 pods instead of 400. Based on what @fredstanley mentioned, it seems like a continuous XDS request  loop is observed, which shouldn't be the case. With the latest code in `main`, I could not repro high CPU usage in a stable setup (no cluster churn). I did observe CPU spikes when app pods are restarted and a full XDS response is being initiated by the controller.

@fredstanley I am unsure how feasible this is for you, but would it be possible to try reproducing this with the latest OSM code from the `main` branch? I can provide the detailed steps required to try this.

**Comment by eduser25** [comment]
I wholeheartedly agree with your assessment, with few caveats.

I wouldn't want to derail @fredstanley's issue, so I'll take it to offline.

**Comment by shashankram** [comment]
> @shashankram We are still on v.0.9.1. It will be hard for us to upgrade to latest. @eduser25 Agree, i also suspect some config is not saved on the proxy and it goes on a viscous cycle of requesting the latest config
> 
> on the envoy proxy i see the version of clusterloadassignment contantly increasing every sec. is it expected to increment that fast without any churn in system ?
> 
> [2021-10-07 00:49:39.500][1][debug][upstream] [source/common/upstream/upstream_impl.cc:279] transport socket match, socket default selected for host with address 10.85.9.211:9070 [2021-10-07 00:49:39.500][1][debug][config] [source/common/config/grpc_subscription_impl.cc:84] gRPC config for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment accepted with 1 resources with version 34 [2021-10-07 00:49:39.500][1][debug][upstream] [source/common/upstream/upstream_impl.cc:279] transport socket match, socket default selected for host with address 10.85.9.244:9052 [2021-10-07 00:49:39.500][1][debug][config] [source/common/config/grpc_subscription_impl.cc:84] gRPC config for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment accepted with 1 resources with version 34 [2021-10-07 00:49:39.500][1][debug][upstream] [source/common/upstream/upstream_impl.cc:279] transport socket match, socket default selected for host with address 10.85.9.248:9067 [2021-10-07 00:49:39.500][1][debug][config] [source/common/config/grpc_subscription_impl.cc:84] gRPC config for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment accepted with 1 resources with version 34

@fredstanley it isn't expected to increase without any churn in the system, so this seems to be the underlying issue. We will look into this and see how to effectively reproduce this.

**Comment by shashankram** [comment]
> Hey @shashankram We dont have any apps that are yet publicly available.
> 
> From my analysis i had a cluster via (cds) without any endpoints for it. And hence the envoys kept discovering endpoints constantly (basically calling eds). After fixing that wrong cluster (created by cds) i no longer see this issue. I assume this might be some wrong configuration on my part. So we can mark this issue closed. I am attaching the pprof pdf output for this issue for future reference. [profile001.pdf](https://github.com/openservicemesh/osm/files/7339801/profile001.pdf)

@fredstanley the high CPU usage due to invalid configuration (cluster without endpoints) was a bug in OSM that was fixed by me. The fix was made in v0.11 version of OSM, so please upgrade to that version.

```
commit 9667f91919f78b89bb049b4b148d1f38abffc4c2
Author: Shashank Ram <21697719+shashankram@users.noreply.github.com>
Date:   Thu Sep 9 11:13:02 2021 -0700

    envoy/eds: respond with endpoints for all services (#4088)
    
    When generating the EDS response for all upstream
    endpoints that the downstream proxy is allowed to
    access, create an EDS response for a service without
    endpoints as well. This is similar to 78455da0 which
    allowed the XDS state machine to converge when there
    are CDS clusters that use EDS for endpoint discovery.
---
commit 78455da0d931e94c8a693cdbf0d4c4e9e7c1cfc1
Author: Shashank Ram <21697719+shashankram@users.noreply.github.com>
Date:   Thu Sep 9 08:46:06 2021 -0700

    envoy/eds: respond to EDS request for cluster with no endpoints (#4085)
    
    It's possible for a cluster created via CDS to have no endpoints,
    such as an apex service referenced in a traffic split resource
    that is not backed by any pods. Respond to the EDS request
    for such a cluster with no endpoints so that Envoy does
    not keep requesting for it.    
```

I am closing this issue since it has already been fixed and upgrading to v0.11 should not result in high CPU usage when a cluster configuration without endpoints exists.
